File: 0051-n-queens/0051-n-queens.cpp
class Solution {
public:
    vector<vector<string>> res;
    vector<vector<char>> board;
    vector<vector<string>> solveNQueens(int n) {
        for (int i = 0; i < n; i++) {
            vector<char> row(n, '.');
            board.push_back(row);
        }
        backtracking(0, n);
        return res;
    }

    void backtracking(int row, int n) {
        if (row == n) {
            res.push_back(convert(board));
            return;
        }

        for (int i = 0; i < n; i++) {
            if (isValid(row, i, n)) {
                board[row][i] = 'Q';
                backtracking(row + 1, n);
                board[row][i] = '.';
            }
        }

    }

    bool isValid(int row, int col, int n) {
        for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        for (int i = row, j = col; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        for (int i = row; i >= 0; i--) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }

        return true;
    }

    vector<string> convert(vector<vector<char>>& board) {
        vector<string> list;
        for (vector<char> chars : board) {
            string str(chars.begin(), chars.end());
            list.push_back(str);
        }
        return list;
    }
};

File: 0051-n-queens/0051-n-queens.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        res = []
        board = [["."] * n for _ in range(n)]

        def solve(row):
            if row == n:
                res.append(["".join(r) for r in board])
                return
            
            for col in range(n):
                if isValid(row, col):
                    board[row][col] = "Q"
                    solve(row + 1)
                    board[row][col] = "."

        def isValid(row, col):
            # check upper
            for i in range(row):
                if board[i][col] == "Q":
                    return False

            # check upper-left

            i, j = row - 1, col - 1

            while i >= 0 and j >= 0:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j -= 1

            i, j = row - 1, col + 1

            while i >= 0 and j < n:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j += 1

            return True

        solve(0)

        return res
        
